File: app/src/main/java/app/baldphone/neo/utils/AppLog.java
package app.baldphone.neo.utils;

import androidx.annotation.NonNull;

import java.util.ArrayDeque;
import java.util.Locale;

/**
 * AppLog – a tiny ring-buffer logger you can use instead of android.util.Log so that recent lines
 * can be added into crash reports.
 *
 * <p>Usage: AppLog.i("Main", "Started"); AppLog.e("Network", "Request failed: " + error);
 *
 * <p>Retrieve recent lines in CrashReporter via AppLog.dumpRecent().
 */
public final class AppLog {
    private static final int MAX_LINES = 100; // keep it small for OOM resilience
    private static final ArrayDeque<String> RING = new ArrayDeque<>(MAX_LINES);

    private AppLog() {}

    public static synchronized void i(String tag, String msg) {
        add("I", tag, msg);
    }

    public static synchronized void w(String tag, String msg) {
        add("W", tag, msg);
    }

    public static synchronized void e(String tag, String msg) {
        add("E", tag, msg);
    }

    public static synchronized void d(String tag, String msg) {
        add("D", tag, msg);
    }

    private static void add(String level, String tag, String msg) {
        long now = System.currentTimeMillis();
        String line =
                String.format(Locale.US, "%tF %tT.%tL %s/%s: %s", now, now, now, level, tag, msg);
        if (RING.size() >= MAX_LINES) RING.removeFirst();
        RING.addLast(line);

        // Also emit to Logcat for developer convenience.
        int logPriority =
                switch (level) {
                    case "E" -> android.util.Log.ERROR;
                    case "W" -> android.util.Log.WARN;
                    case "D" -> android.util.Log.DEBUG;
                    default -> android.util.Log.INFO;
                };
        android.util.Log.println(logPriority, tag, msg);
    }

    @NonNull
    public static synchronized String dumpRecent() {
        StringBuilder sb = new StringBuilder();
        for (String s : RING) sb.append(s).append('\n');
        return sb.toString();
    }
}

File: app/src/main/java/app/baldphone/neo/utils/AppLog.kt
package app.baldphone.neo.utils

import android.util.Log

import com.bald.uriah.baldphone.BuildConfig

import java.util.concurrent.atomic.AtomicInteger

/**
 * AppLog – Optimized logging utility.
 * Uses inline functions to avoid string allocation when logging is disabled.
 */
object AppLog {
    private const val MAX_LINES = 200
    private val buffer = arrayOfNulls<LogEntry>(MAX_LINES)
    private val index = AtomicInteger(0)

    private data class LogEntry(
        val timestamp: Long, val level: Int, val tag: String, val msg: String
    )

    inline fun v(tag: String, msg: () -> String) = logIfEnabled(Log.VERBOSE, tag, msg)
    inline fun d(tag: String, msg: () -> String) = logIfEnabled(Log.DEBUG, tag, msg)
    inline fun i(tag: String, msg: () -> String) = logIfEnabled(Log.INFO, tag, msg)
    inline fun w(tag: String, msg: () -> String) = logIfEnabled(Log.WARN, tag, msg)
    inline fun e(tag: String, msg: () -> String) = logIfEnabled(Log.ERROR, tag, msg)

    /** Overloads for Throwable */
    fun w(t: Throwable?, msg: String? = null) = add(Log.WARN, "AppLog", msg ?: "", t)
    fun e(t: Throwable?, msg: String? = null) = add(Log.ERROR, "AppLog", msg ?: "", t)

    /** Backward compatibility overloads */
    @JvmStatic
    fun v(tag: String, msg: String) = add(Log.VERBOSE, tag, msg, null)

    @JvmStatic
    fun d(tag: String, msg: String) = add(Log.DEBUG, tag, msg, null)

    @JvmStatic
    fun i(tag: String, msg: String) = add(Log.INFO, tag, msg, null)

    @JvmStatic
    fun w(tag: String, msg: String) = add(Log.WARN, tag, msg, null)

    @JvmStatic
    fun e(tag: String, msg: String) = add(Log.ERROR, tag, msg, null)

    @PublishedApi
    internal inline fun logIfEnabled(level: Int, tag: String, msg: () -> String) {
        if (BuildConfig.DEBUG || level >= Log.WARN) {
            add(level, tag, msg(), null)
        }
    }

    @PublishedApi
    internal fun add(level: Int, tag: String, message: String, t: Throwable?) {
        // Keep for Java users
        if (!BuildConfig.DEBUG && level < Log.WARN) return

        val fullMsg = if (t != null) "$message\n${Log.getStackTraceString(t)}" else message

        // System Log
        Log.println(level, tag, fullMsg)

        // Ring Buffer
        val ts = System.currentTimeMillis()
        val i = (index.getAndIncrement() and Int.MAX_VALUE) % MAX_LINES
        buffer[i] = LogEntry(ts, level, tag, fullMsg)
    }

    @JvmStatic
    fun dumpRecent(): String {
        val sb = StringBuilder()
        val currentIndex = index.get()
        val count = currentIndex.coerceAtMost(MAX_LINES)
        val start =
            if (currentIndex < MAX_LINES) 0 else (currentIndex and Int.MAX_VALUE) % MAX_LINES

        for (i in 0 until count) {
            val pos = (start + i) % MAX_LINES
            val entry = buffer[pos] ?: continue
            val ts = entry.timestamp.toFullDateTimeString(includeMillis = true)
            sb.appendLine("$ts ${levelToString(entry.level)}/${entry.tag}: ${entry.msg}")
        }
        return sb.toString()
    }

    private fun levelToString(level: Int) = when (level) {
        Log.VERBOSE -> "V"
        Log.DEBUG -> "D"
        Log.INFO -> "I"
        Log.WARN -> "W"
        Log.ERROR -> "E"
        else -> "?"
    }
}
